### PR TITLE
[Geospatial] Add new geo shape filter field to support geospatial search query

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Add new filter query, 'geo_shape' to search geospatial field types . This is required as part of Maps plugin which
uses geo_shape filter query to search documents.
 
### Issues Resolved
https://github.com/opensearch-project/dashboards-maps/issues/213
 
### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  };
+};
+
+export type GeoShapeFilter = Filter & {
+  meta: GeoShapeFilterMeta;
+  geo_shape: any;
+};
+
+export const isGeoShapeFilter = (filter: any): filter is GeoShapeFilter =>
+  filter && filter.geo_shape;
```
Comment (AMoo-Miki) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:45]:
```suggestion
  filter?.geo_shape;
```

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+export const getGeoShapeFilterField = (filter: GeoShapeFilter) => {
+  return filter.geo_shape && Object.keys(filter.geo_shape).find((key) => key !== 'ignore_unmapped');
```
Comment (AMoo-Miki) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:48]: This would be safer with the extra check
```suggestion
  return filter?.geo_shape && Object.keys(filter.geo_shape).find((key) => key !== 'ignore_unmapped');
```

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  type: 'Polygon';
+  coordinates: Position[][];
+}
+
+export interface MultiPolygon {
+  type: 'MultiPolygon';
+  coordinates: Position[][][];
+}
+
+// Will support polygon & multi polygon at this moment
```
Comment (AMoo-Miki) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:28]: Do we need a "ToDo" to potentially add others?

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 1 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:45]: Ack

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 2 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:48]: Here geo_shape is not an optional field for GeoShapeFilter, adding chaining is not required.

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 3 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:28]: Sure. Does adding TODO is enough or do you recommend github issue too?

**Comment by VijayanB** [comment]
@AMoo-Miki Can you take a look at PR one more time? Thanks. Sorry for the rush, since we will be building feature in Maps using this PR and this is blocker for our current development.

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 3 above)
Comment (AMoo-Miki) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts]: You know what? GH issue would be the best :)

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 2 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts]: updated

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 3 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts]: Added https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3612

**Comment by kavilla** [comment]
@VijayanB for speed could you provide a screen shot or describe recreation steps how to be able to use this new functionality.

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  params: {
+    shape?: ShapeFilter;
+    indexed_shape?: PreIndexedShapeFilter;
+    relation?: GeoShapeRelation;
+  };
+};
+
+export type GeoShapeFilter = Filter & {
+  meta: GeoShapeFilterMeta;
+  geo_shape: any;
```
Comment (ananzh) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:41]: Will we consider define a more specific interface or type for the geo_shape property?

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+};
+
+export const isGeoShapeFilter = (filter: any): filter is GeoShapeFilter => filter?.geo_shape;
+
+export const getGeoShapeFilterField = (filter: GeoShapeFilter) => {
+  return (
```
Comment (ananzh) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:47]: NIT: should we explicitly define `return undefined` to make it more clearer?

```
export const getGeoShapeFilterField = (filter: GeoShapeFilter) => {
  if (filter.geo_shape === undefined) {
    return undefined;
  }
  return Object.keys(filter.geo_shape).find((key) => key !== 'ignore_unmapped');
};
```

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.test.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          disabled: false,
+          negate: false,
+          alias: null,
+          params: geoShapeQuery,
+        },
+      };
+      const result = getGeoShapeFilterField(filter);
+      expect(result).toBe('geoPointField');
+    });
+  });
```
Comment (ananzh) [inline src/plugins/data/common/opensearch_query/filters/geo_shape_filter.test.ts:51]: I think we should also add a simple edge case unit test for `filter.geo_shape === undefined` since you have the logic `filter?.geo_shape && Object.keys(filter.geo_shape).find((key) => key !== 'ignore_unmapped' ` in the `getGeoShapeFilterField` function. Maybe something like:


```
it('should return undefined if filter.geo_shape is undefined', () => {
    const filter: GeoShapeFilter = {
      geo_shape: undefined,
      meta: {
        disabled: false,
        negate: false,
        alias: null,
        params: {
          shape: undefined,
          relation: 'intersects',
        },
      },
    };
    const result = getGeoShapeFilterField(filter);
    expect(result).toBeUndefined();
  });
```

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 4 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:41]: geo_shape will have dynamic field name and "ignored_unmapped" as well. I couldn't find a way to define both in same structure. Do you know how to do that?

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 5 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:50]: Updated

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.test.ts (hunk 6 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.test.ts:66]: Added new test

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 4 above)
Comment (ananzh) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:41]: Not quite sure about what should be geo_shape, but it is similar to meta, will this work?

```
export type GeoShapeFilter = Filter & {
  meta: GeoShapeFilterMeta;
  geo_shape: {
    [fieldName: string]: {
      shape?: ShapeFilter;
      indexed_shape?: PreIndexedShapeFilter;
      relation?: GeoShapeRelation;
      ignore_unmapped?: boolean;
    };
  };
};

```

**Comment by VijayanB** [comment]
> @VijayanB for speed could you provide a screen shot or describe recreation steps how to be able to use this new functionality.

Sure. geo_shape is a query type that can be used to define search query. Currently Region Maps and Coordinate maps uses  geo_polygon query type, but, it is deprecated and not encourage any more. Hence, this PR, will understand any query with Filter of type GeoShape and can serialize into geo_shape query for search request . 

Maps will be using geo_shape filter instead of geo_polygon.

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 4 above)
Comment (ananzh) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:41]: Not a blocker. I will just approve it.

**Review by ananzh — DISMISSED** [review]

File: src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts (hunk 4 above)
Comment (VijayanB) [reply src/plugins/data/common/opensearch_query/filters/geo_shape_filter.ts:41]: Here ignore_unmapped should be inside geo_shape , something like below
```
geo_shape: {
    [fieldName: string]: {
      shape?: ShapeFilter;
      indexed_shape?: PreIndexedShapeFilter;
      relation?: GeoShapeRelation;
    },
   ignore_unmapped?: boolean;
}
```
but typescript don't like this, since ignore_unmapped is also string and it can't differentiate whether is it part of fieldName or ignore_mapped.

**Review by ananzh — DISMISSED** [review]

**Comment by VijayanB** [comment]
@kavilla @AMoo-Miki Can you please take a look at this PR?

**Review by ananzh — APPROVED** [review]

**Review by kavilla — APPROVED** [review]
